### ISSUE TITLE
Diagnostics and visualization might show incorrect dependencies for conditional registrations

### ISSUE THREAD

**Issue by dotnetjunkie** [simpleinjector/SimpleInjector#349]
Simple Injector's diagnostic subsystem uses the list of `KnownRelationship` instances that is provided by the `Registration` instance. This list is than used to run diagnostics on the configuration. This same list is used by the graph visualization feature.

As a simplification, the `Registration` class caches the list of `KnownRelationship` instances while it generates an `Expression` for the given registration. Caching however becomes a problem when `Registration` instances are reused by multiple `InstanceProducer` instancess in combination with a dependency that is conditional. In case each `InstanceProducer` gets a different dependency, the `Registration` holds both dependencies in its `KnownRelationship` collection. This causes the diagnostic system and graph visualization to see both dependencies for both `InstanceProducer`s.

The following example demonstrates the problem:

``` c#
public interface IX { }
public interface IY { }
public interface ILogger { }
public class XLogger : ILogger { }
public class YLogger : ILogger { }
public class A : IX, IY { public A(ILogger logger) { } }

private static void Main(string[] args)
{
    var container = new Container();
    container.Options.DefaultScopedLifestyle = new ThreadScopedLifestyle();

    container.RegisterConditional(typeof(ILogger), typeof(XLogger), Lifestyle.Singleton,
        c => c.Consumer.ServiceType == typeof(IX));
    container.RegisterConditional(typeof(ILogger), typeof(YLogger), Lifestyle.Singleton,
        c => c.Consumer.ServiceType == typeof(IY));

    Registration reg = Lifestyle.Transient.CreateRegistration<A>(container);

    container.AddRegistration(typeof(IX), reg);
    container.AddRegistration(typeof(IY), reg);

    container.Verify();

    Console.WriteLine(container.GetRegistration(typeof(IX)).VisualizeObjectGraph());
    Console.WriteLine(container.GetRegistration(typeof(IY)).VisualizeObjectGraph());
}
```

The application is expected to output:

```
A(
    XlLogger())
A(
    YLogger())
```

But instead, the output is:

```
A(
    XlLogger(),
    YLogger())
A(
    XlLogger(),
    YLogger())
```

The solution is to stop caching `KnownRelationship` instances inside the `Registration` class, but this does mean we need to introduce a breaking change, because now the `Registration` should return that list from its `BuildExpression` method.

**Comment by dotnetjunkie** [comment]
This requires a total redesign of the `Registration` base class (which might be a good thing). The `Registration` class is now mutable and statefull, but it would be better to go with an immutable/stateless model where it builds up all the information when requested and directly forgets about this information after it returned that data.

**Comment by dotnetjunkie** [comment]
With #348 the `Consumer.ServiceType` is removed which is the cause of this issue to rise. The fix for #348 also fixes this issue.
